src/contracts/middleware/StakeRegistry.sol

```diff
@@ -0,0 +1,461 @@
+
+        // add the strategies considered and multipliers for each quorum
+        for (uint8 quorumNumber = 0; quorumNumber < _quorumStrategiesConsideredAndMultipliers.length;) {
+            _setMinimumStakeForQuorum(quorumNumber, _minimumStakeForQuorum[quorumNumber]);
+            _createQuorum(_quorumStrategiesConsideredAndMultipliers[quorumNumber]);
+            unchecked {
+                ++quorumNumber;
+            }
+        }
+    }
+
+    /**
+     * @notice Returns the `index`-th entry in the `operatorIdToStakeHistory[operatorId][quorumNumber]` array.
+     * @param quorumNumber The quorum number to get the stake for.
+     * @param operatorId The id of the operator of interest.
+     * @param index Array index for lookup, within the dynamic array `operatorIdToStakeHistory[operatorId][quorumNumber]`.
+     * @dev Function will revert if `index` is out-of-bounds.
+     */
+    function getStakeUpdateForQuorumFromOperatorIdAndIndex(uint8 quorumNumber, bytes32 operatorId, uint256 index)
+        external
+        view
+        returns (OperatorStakeUpdate memory)
+    {
+        return operatorIdToStakeHistory[operatorId][quorumNumber][index];
+    }
+
+    /**
+     * @notice Returns the `index`-th entry in the dynamic array of total stake, `_totalStakeHistory` for quorum `quorumNumber`.
+     * @param quorumNumber The quorum number to get the stake for.
+     * @param index Array index for lookup, within the dynamic array `_totalStakeHistory[quorumNumber]`.
+     */
+    function getTotalStakeUpdateForQuorumFromIndex(uint8 quorumNumber, uint256 index) external view returns (OperatorStakeUpdate memory) {
+        return _totalStakeHistory[quorumNumber][index];
+    }
+
+    /// @notice Returns the indices of the operator stakes for the provided `quorumNumber` at the given `blockNumber`
+    function getStakeUpdateIndexForOperatorIdForQuorumAtBlockNumber(bytes32 operatorId, uint8 quorumNumber, uint32 blockNumber)
+        external
+        view
+        returns (uint32)
+    {
+        return _getStakeUpdateIndexForOperatorIdForQuorumAtBlockNumber(operatorId, quorumNumber, blockNumber);
+    }
+
+
+    /// @notice Returns the indices of the total stakes for the provided `quorumNumbers` at the given `blockNumber`
+    function getTotalStakeIndicesByQuorumNumbersAtBlockNumber(uint32 blockNumber, bytes calldata quorumNumbers) external view returns(uint32[] memory) {
+        uint32[] memory indices = new uint32[](quorumNumbers.length);
+        for (uint256 i = 0; i < quorumNumbers.length; i++) {
+            uint8 quorumNumber = uint8(quorumNumbers[i]);
+            uint32 length = uint32(_totalStakeHistory[quorumNumber].length);
+            for (uint32 j = 0; j < length; j++) {
+                if (_totalStakeHistory[quorumNumber][length - j - 1].updateBlockNumber <= blockNumber) {
+                    indices[i] = length - j - 1;
+                    break;
+                }
+            }
+        }
+        return indices;
+    }
```

Comment:
what's the intended failure mode of this function? e.g. if I input a block number that's too-low (or too-high)?

Reply:
seems like this will fail with underflow rn (?) on this line:
`indices[i] = length - j - 1;`

Reply:
> seems like this will fail with underflow rn (?) on this line:

I'm wrong, it'll do `(length - (length - 1) - 1) = 0`

so I guess the behavior here is returning index 0 if the block number is from before the quorum existed. seems fine, but intention should still be documented

Reply:
talked about this offline. seems like the desired behavior here is _reversion_. We should document and implement this.

Reply:
I'll look and see if there are other similar instances.

Reply:
> I'll look and see if there are other similar instances.

AFAICT there aren't other instances of similar quirky edge-case behavior in this PR. Similar functions seem to all either revert in the same kind of scenario, or return a special index + explain (in comments) that they will return a special index in these edge cases.